### PR TITLE
docs(LightningModule): update docs for `.training` mode in loops

### PR DESCRIPTION
**Pull request by adosar**
Related to https://github.com/Lightning-AI/pytorch-lightning/pull/18951#issuecomment-2633423417.

Update the pseudocode of validation loop according to #18951:

> When the validation loop ends, and before switching to training, it restores the `.training mode` on all submodules to what it was before.

and add a corresponding note to `{validation,test,predict}_step` since they exhibit this behavior as can be seen in the following snippet:
```python
from lightning.pytorch.demos.boring_classes import BoringModel
import lightning as L
import warnings

warnings.filterwarnings('ignore')

trainer = L.Trainer(max_epochs=1)
loop = trainer.test

litmodel = BoringModel()
litmodel.train()
print('Before loop', litmodel.training)
loop(litmodel)
print('After loop', litmodel.training)

litmodel = BoringModel()
litmodel.eval()
print('Before loop', litmodel.training)
loop(litmodel)
print('After loop', litmodel.training)
```

```
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
Before loop True
Testing DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 64/64 [00:00<00:00, 1299.38it/s]
After loop True
Before loop False
Testing DataLoader 0: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 64/64 [00:00<00:00, 1332.01it/s]
After loop False
```
@awaelchli Can you please confirm that this is the intended (default) behavior of the loops?

Additional changes:
* Fix incorrect comment in `lightning_module.rst` that `trainer.test(model)` loads the best weights. According to the [docs](https://lightning.ai/docs/pytorch/stable/common/trainer.html#test), If `ckpt_path=None` and the model instance was passed, use the current weights.

## What does this PR do?

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by adosar** [comment]
cc @Borda

File: docs/source-pytorch/common/lightning_module.rst [hunk 1]
```diff
@@ -286,6 +286,9 @@ Under the hood, Lightning does the following (pseudocode):
         # ...
 
         if validate_at_some_point:
+            # capture .training mode of every submodule
+            capture_training_mode()
```
Comment (Borda) [inline docs/source-pytorch/common/lightning_module.rst]: Can we add some examples of what the `capture_training_mode` and later `restore_training_mode` us used for? Now it's a bit confusing since you declare/use a cute function but a new function...

**Review by Borda — CHANGES_REQUESTED** [review]

File: docs/source-pytorch/common/lightning_module.rst (hunk 1 above)
Comment (adosar) [reply docs/source-pytorch/common/lightning_module.rst]: @Borda I am thinking of adding a note in the docs instead of introducing these functions in the pseudocode of the validation loop. The take home message should be that Lightning takes care of ensuring that layers set in eval mode by user remain in eval mode.

Comment (SkafteNicki) [reply docs/source-pytorch/common/lightning_module.rst]:
```suggestion
            training_mode = model.training
```
since this is just pseudocode, lets just use standard pytorch notation

File: docs/source-pytorch/common/lightning_module.rst [hunk 2]
```diff
@@ -295,9 +298,11 @@ Under the hood, Lightning does the following (pseudocode):
                 val_out = model.validation_step(val_batch, val_batch_idx)
             # ----------------- VAL LOOP ---------------
 
-            # enable grads + batchnorm + dropout
+            # enable grads
             torch.set_grad_enabled(True)
-            model.train()
+
+            # restore .training mode of every submodule
+            restore_training_mode()
```
Comment (SkafteNicki) [inline docs/source-pytorch/common/lightning_module.rst]:
```suggestion
            model.training = training_mode
```

File: src/lightning/pytorch/core/module.py [hunk 3]
```diff
@@ -814,9 +814,10 @@ def validation_step(self, batch, batch_idx, dataloader_idx=0):
             If you don't need to validate you don't need to implement this method.
 
         Note:
-            When the :meth:`validation_step` is called, the model has been put in eval mode
-            and PyTorch gradients have been disabled. At the end of validation,
-            the model goes back to training mode and gradients are enabled.
+            When the :meth:`validation_step` is called, the model has been put
```
Comment (SkafteNicki) [inline src/lightning/pytorch/core/module.py:846]: for completeness can we add a inverse note to the training step, e.g. that when `train_step` is called, the model has NOT been put into train mode and that it is something the user should be aware of

File: src/lightning/pytorch/core/module.py [hunk 4]
```diff
@@ -760,6 +760,10 @@ def training_step(self, batch, batch_idx):
             When ``accumulate_grad_batches`` > 1, the loss returned here will be automatically
             normalized by ``accumulate_grad_batches`` internally.
 
+        Note:
+            When the :meth:`training_step` is called, the training/eval mode of the model and submodules is
+            whatever it was when provided to `trainer.fit`.
+
```
Comment (adosar) [inline src/lightning/pytorch/core/module.py]: Changing `trainer.fit` to ``` :meth:`.Trainer.fit` ``` so it creates  a link?

Comment (SkafteNicki) [reply src/lightning/pytorch/core/module.py]:
```suggestion
        Note:
            When the :meth:`training_step` is called, the training/eval mode of the model and submodules is
            whatever it was when provided to :meth:`~lightning.pytorch.trainer.trainer.Trainer.fit`.

```
